pkg/spdk/replica.go

```diff
@@ -1543,6 +1609,558 @@ func (r *Replica) SnapshotRangeHashGet(spdkClient *spdkclient.Client, snapshotNa
+		aggregatedErrors = append(aggregatedErrors, err)
+	} else {
+		r.snapshotCloningDstCache.cloningLvol = nil
+	}
+
+	if _, err := spdkClient.BdevLvolDelete(spdktypes.GetLvolAlias(r.LvsName, cloningLvolName)); err != nil && !jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
+		r.log.WithError(err).Errorf("Failed to delete the cloning lvol %s for cloning dst cleanup", cloningLvolName)
+		aggregatedErrors = append(aggregatedErrors, err)
+	} else {
+		r.snapshotCloningDstCache.cloningLvol = nil
+	}
+
+	tmpSnapName := GetTmpSnapNameForCloningLvol(r.Name)
+	if _, err := spdkClient.BdevLvolDelete(spdktypes.GetLvolAlias(r.LvsName, tmpSnapName)); err != nil && !jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
+		r.log.WithError(err).Errorf("Failed to delete the tmp snapshot %s for cloning dst cleanup", cloningLvolName)
+		aggregatedErrors = append(aggregatedErrors, err)
+	}
+
+	r.snapshotCloningDstCache.srcReplicaName = ""
+	r.snapshotCloningDstCache.srcReplicaAddress = ""
+	if r.snapshotCloningDstCache.monitorCancelFunc != nil {
+		r.snapshotCloningDstCache.monitorCancelFunc()
+		r.snapshotCloningDstCache.monitorCancelFunc = nil
+	}
+
+	if clearStatus {
+		r.snapshotCloningDstCache.processedClusters = 0
+		r.snapshotCloningDstCache.totalClusters = 0
+		r.snapshotCloningDstCache.cloningError = ""
+		r.snapshotCloningDstCache.cloningState = ""
+	}
+
+	return util.CombineErrors(aggregatedErrors...)
+}
+
+func (r *Replica) snapshotLinkedCloneSrcStart(spdkClient *spdkclient.Client, snapshotName, dstReplicaName string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to do snapshotLinkedCloneSrcStart")
+	}()
+
+	bdevLvolMap, err := GetBdevLvolMapWithFilter(spdkClient, r.replicaLvolFilter)
+	if err != nil {
+		return err
+	}
+
+	existingParentOfDstReplica := ""
+	for lvolName, lvol := range bdevLvolMap {
+		if types.IsBackingImageSnapLvolName(lvolName) {
+			continue
+		}
+		for _, childLvolName := range lvol.DriverSpecific.Lvol.Clones {
+			if childLvolName == dstReplicaName {
+				existingParentOfDstReplica = lvolName
+				continue
+			}
+			if !IsReplicaLvol(r.Name, childLvolName) {
+				return fmt.Errorf("there are already another linked-clone lvol %v in src replica %v. "+
+					"Each src replica can only has 1 linked-clone lvol at a time", childLvolName, r.Name)
+			}
+		}
+	}
+
+	snapLvolName := GetReplicaSnapshotLvolName(r.Name, snapshotName)
+	snapLvol := r.SnapshotLvolMap[snapLvolName]
+	if snapLvol == nil {
+		return fmt.Errorf("cannot find snapshot %s for src replica %s", snapshotName, r.Name)
+	}
+
+	if existingParentOfDstReplica != "" {
+		if existingParentOfDstReplica != snapLvolName {
+			return fmt.Errorf("dst replica already has a different parent %q than the snapshot %v", dstReplicaName, snapshotName)
+		}
+		// Operation is already satisfied
+		return nil
+	}
+
+	set, err := spdkClient.BdevLvolSetParent(spdktypes.GetLvolAlias(r.LvsName, dstReplicaName), snapLvol.Alias)
+	if err != nil {
+		return err
+	}
```

Comment:
If the src replica and dst replica are on the same node but in different lvs disks, does it still work?